pkg/karmadactl/util/check.go

```diff
@@ -6,72 +6,138 @@ import (
 }
 
 // WaitForDeploymentRollout  wait for Deployment reaches the ready state or timeout.
-func WaitForDeploymentRollout(c kubernetes.Interface, dep *appsv1.Deployment, timeoutSeconds int) error {
-	var lastErr error
-	pollError := wait.PollImmediate(time.Second, time.Duration(timeoutSeconds)*time.Second, func() (bool, error) {
-		d, err := c.AppsV1().Deployments(dep.GetNamespace()).Get(context.TODO(), dep.GetName(), metav1.GetOptions{})
-		if err != nil {
-			lastErr = err
-			return false, nil
-		}
-		if d.Generation != d.Status.ObservedGeneration {
-			lastErr = fmt.Errorf("current generation %d, observed generation %d",
-				d.Generation, d.Status.ObservedGeneration)
-			return false, nil
-		}
-		if (d.Spec.Replicas != nil) && (d.Status.UpdatedReplicas < *d.Spec.Replicas) {
-			lastErr = fmt.Errorf("the number of pods targeted by the deployment (%d pods) is different "+
-				"from the number of pods targeted by the deployment that have the desired template spec (%d pods)",
-				*d.Spec.Replicas, d.Status.UpdatedReplicas)
-			return false, nil
+func WaitForDeploymentRollout(c kubernetes.Interface, dep *appsv1.Deployment, timeoutSeconds time.Duration) error {
+	depWatcher, err := c.AppsV1().Deployments(dep.GetNamespace()).Watch(context.Background(), metav1.SingleObject(dep.ObjectMeta))
+	if err != nil {
+		return err
+	}
+
+	for {
+		select {
+		case <-time.After(timeoutSeconds * time.Second):
+			klog.Errorf("wait for Deployment(%s/%s) rollout: timeout", dep.GetNamespace(), dep.GetName())
+			return nil
+		case event, ok := <-depWatcher.ResultChan():
+			if !ok {
+				return fmt.Errorf("wait for Deployment(%s/%s) rollout: channel broken", dep.GetNamespace(), dep.GetName())
+			}
+
+			status, err := deploymentStatus(event.Object)
+			if err != nil {
+				return err
+			} else if status == StatusReady {
+				return nil
+			}
 		}
-		if d.Status.AvailableReplicas < d.Status.UpdatedReplicas {
-			lastErr = fmt.Errorf("expected %d replicas, got %d available replicas",
-				d.Status.UpdatedReplicas, d.Status.AvailableReplicas)
-			return false, nil
+	}
+}
+
+// Reference https://github.com/kubernetes-sigs/application/blob/e5329b1b083ece8abb4b4efaf738ab9277fbb2ce/controllers/status.go#L78-#L92
+func stsStatus(o runtime.Object) (string, error) {
+	u, err := runtime.DefaultUnstructuredConverter.ToUnstructured(o)
+	if err != nil {
+		return StatusUnknown, err
+	}
+
+	sts := &appsv1.StatefulSet{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u, sts); err != nil {
+		return StatusUnknown, err
+	}
+
+	if sts.Status.ObservedGeneration == sts.Generation &&
+		sts.Status.Replicas == *sts.Spec.Replicas &&
+		sts.Status.ReadyReplicas == *sts.Spec.Replicas &&
+		sts.Status.CurrentReplicas == *sts.Spec.Replicas {
+		return StatusReady, nil
+	}
+	return StatusInProgress, nil
+}
+
+// Reference https://github.com/kubernetes-sigs/application/blob/e5329b1b083ece8abb4b4efaf738ab9277fbb2ce/controllers/status.go#L94-#L132
+//
+//nolint:gocyclo
```

Comment:
Personally, I'm hesitant to introduce `nolint`. That's my concern.

Reply:
This is my reference to the kubernetes-sigs/application code, only adding ToUnstructured, probably because the logic of if is more complicated to determine